daml-lf/archive/src/main/scala/com/digitalasset/daml/lf/archive/DarReader.scala

```diff
@@ -21,101 +23,127 @@ class DarReader[A](
       extends RuntimeException(
-        s"Invalid zip entry: ${zipEntry.getName: String}, DAR: ${darInfo(darFile): String}")
+        s"Invalid zip entryName: ${name: String}, DAR: ${darInfo(entries): String}")
 
-  case class InvalidLegacyDar(darFile: ZipFile)
-      extends RuntimeException(s"Invalid DAR: ${darInfo(darFile): String}")
+  case class InvalidLegacyDar(entries: ZipEntries)
+      extends RuntimeException(s"Invalid Legacy DAR: ${darInfo(entries)}")
 
-  private def darInfo(darFile: ZipFile): String =
-    s"${darFile.getName: String}, content: [${darFileNames(darFile).mkString(", "): String}}]"
+  private def darInfo(entries: ZipEntries): String =
+    s"${entries.name}, content: [${darFileNames(entries).mkString(", "): String}}]"
 
-  private def darFileNames(darFile: ZipFile): List[String] =
-    darFile.entries.asScala.toList.map(_.getName)
+  private def darFileNames(entries: ZipEntries): Iterable[String] =
+    entries.entries.keys
 }
 
 object DarReader {
+
+  private val ManifestName = "META-INF/MANIFEST.MF"
+
+  private[archive] case class ZipEntries(name: String, entries: Map[String, (Long, InputStream)]) {
+
+    def getInputStreamFor(entryName: String): Try[(Long, InputStream)] = {
+      entries.get(entryName) match {
+        case Some((size, is)) => Success(size -> is)
+        case None => Failure(InvalidZipEntry(entryName, this))
+      }
+    }
+
+    def readDalfNames(
+        readDalfNamesFromManifest: InputStream => Try[Dar[String]]): Try[Dar[String]] =
+      parseDalfNamesFromManifest(readDalfNamesFromManifest).recoverWith {
+        case NonFatal(e1) =>
+          findLegacyDalfNames().recoverWith {
+            case NonFatal(_) => Failure(InvalidDar(this, e1))
+          }
+      }
+
+    private def parseDalfNamesFromManifest(
+        readDalfNamesFromManifest: InputStream => Try[Dar[String]]): Try[Dar[String]] =
+      bracket(getInputStreamFor(ManifestName)) { case (_, is) => close(is) }
+        .flatMap { case (_, is) => readDalfNamesFromManifest(is) }
+
+    // There are three cases:
+    // 1. if it's only one .dalf, then that's the main one
+    // 2. if it's two .dalfs, where one of them has -prim in the name, the one without -prim is the main dalf.
+    // 3. parse error in all other cases
+    private def findLegacyDalfNames(): Try[Dar[String]] = {
+      val dalfs: List[String] = entries.keys.filter(isDalf).toList
+
+      dalfs.partition(isPrimDalf) match {
+        case (List(prim), List(main)) => Success(Dar(main, List(prim)))
+        case (List(prim), Nil) => Success(Dar(prim, List.empty))
+        case (Nil, List(main)) => Success(Dar(main, List.empty))
+        case _ => Failure(InvalidLegacyDar(this))
+      }
+    }
+
+    private def isDalf(s: String): Boolean = s.toLowerCase.endsWith(".dalf")
+
+    private def isPrimDalf(s: String): Boolean = s.toLowerCase.contains("-prim") && isDalf(s)
+  }
+
+  private[archive] def close(is: InputStream): Try[Unit] = Try(is.close())
```

Comment:
this `close` is not used anywhere. Let's delete it or change it so it closes `Try[(Long, InputStream)]`

Reply:
it's actually used in `parseDalfNamesFromManifest` and `parseOne`. Nevertheless I find it to be an overkill so I will just inline them.